drivers/console/aarch64/console.S

```diff
@@ -3,103 +3,240 @@
+
+#if !ERROR_DEPRECATED
+	/* Set up compatiblity symbols for platforms using old console API. */
+	declare_console core
+
+	.globl	console_init
+	/* -----------------------------------------------
 	 * int console_init(uintptr_t base_addr,
 	 * unsigned int uart_clk, unsigned int baud_rate)
-	 * Function to initialize the console without a
-	 * C Runtime to print debug information. It saves
-	 * the console base to the data section.
+	 * Deprecated legacy function for initializing a
+	 * single console. Current platforms should call
+	 * console_<driver>_register for the relevant
+	 * driver instead. This needs to clear the console
+	 * list first because some legacy platforms called
+	 * console_init more than once and expect it to
+	 * overwrite previous console state.
 	 * In: x0 - console base address
 	 *     w1 - Uart clock in Hz
 	 *     w2 - Baud rate
-	 * out: return 1 on success else 0 on error
-	 * Clobber list : x1 - x4
+	 * Clobber list : x0 - x4, x14, x15
 	 * -----------------------------------------------
 	 */
 func console_init
-	/* Check the input base address */
-	cbz	x0, init_fail
-	adrp	x3, console_base
-	str	x0, [x3, :lo12:console_base]
-	b	console_core_init
-init_fail:
-	ret
+	adrp	x14, console_list
+	str	xzr, [x14, :lo12:console_list]
+	b	console_core_register
 endfunc console_init
+#endif
+
+	/* -----------------------------------------------
+	 * console_info_t *console_unregister(
+	 *                        console_info_t *console)
+	 * Function to find a specific console in the list
+	 * of currently active consoles and remove it.
+	 * Should usually only be called by
+	 * declare_console expansions.
+	 * In: x0 - address of console structure to remove
+	 * Out: x0 - removed address, or NULL if not found
+	 * Clobber list: x0, x1, x14
+	 * -----------------------------------------------
+	 */
+func console_unregister
+	adrp	x14, console_list
+	add	x14, x14, :lo12:console_list	/* X14 = ptr to first struct */
+	ldr	x1, [x14]			/* X3 = first struct */
+
+unregister_loop:
+	cbz	x1, unregister_not_found
+	cmp	x0, x1
+	b.eq	unregister_found
+	ldr	x14, [x14]			/* X14 = next ptr of struct */
+	ldr	x1, [x14]			/* X1 = next struct */
+	b	unregister_loop
+
+unregister_found:
+	ldr	x1, [x1]			/* X1 = next struct */
+	str	x1, [x14]			/* prev->next = cur->next */
+	ret
+
+unregister_not_found:
+	mov	x0, #0				/* return NULL if not found */
+	ret
+endfunc console_unregister
 
 	/* -----------------------------------------------
 	 * void console_uninit(void)
-	 * Function to finish the use of console driver.
-	 * It sets the console_base as NULL so that any
-	 * further invocation of `console_putc` or
-	 * `console_getc` APIs would return error.
+	 * Function to switch from boot to runtime state.
+	 * The name is misleading for compatibility with
+	 * the legacy API -- this used to uninitialize all
+	 * consoles, but now it just quiesces consoles
+	 * that are not marked as "runtime".
+	 * Clobber list: x0, x1
```

Comment:
OK, now I understand the behaviour you're expecting. This is different to the behaviour expected by the platforms I know about. For the latter, we expect there to be 2 serial consoles, one for boot and one for runtime. The boot console is handed over to other boot components after TF boot (normal world firmware and eventually Linux). On those platforms, the crash console is currently statically configured to be the same as the runtime console.

As far as I can see, both our use-cases can be implemented by both our suggestions, although yours requires additional state. To implement your expected behaviour with my suggestion, the platform would override bl31_plat_runtime_setup() to disable the serial boot console and leave the runtime memory console active, and also override plat_crash_console_init to re-enable the serial console for crashes (or enable it the first time for early crashes).

The advantages to your suggestion that I can see, are a) there's generally no need to override plat_crash_console_init(), unless you want console output for early crashes, and b) some platforms can do all console registration/enablement in one place (platform setup). a) seems quite weak as overriding this function is easy and early crash debugging is generally important. Also, platforms that do override plat_crash_console_init will need to remember to call console_start_crash to keep things working. b) doesn't help the platforms I mentioned before, as it's only possible to have one serial console registered at a time, so the platforms would need to unregister/register the console anyway during the runtime switch.

Finally, even if we do retain this console state, this shouldn't change the implementation of console_uninit() in this unintuitive way; it should change the plat/common implementation of bl31_plat_runtime_setup().